File: backend/app/cache.py
"""Tiny thread-safe in-memory TTL cache.

Kept dependency-free. For a single-instance deployment this is sufficient;
swap for Redis if the app is scaled horizontally.
"""

from __future__ import annotations

import threading
import time
from typing import Generic, TypeVar

T = TypeVar("T")


class TTLCache(Generic[T]):
    def __init__(self, ttl_seconds: int) -> None:
        self._ttl = ttl_seconds
        self._store: dict[str, tuple[float, T]] = {}
        self._lock = threading.Lock()

    def get(self, key: str) -> T | None:
        with self._lock:
            entry = self._store.get(key)
            if entry is None:
                return None
            expires_at, value = entry
            if time.time() > expires_at:
                self._store.pop(key, None)
                return None
            return value

    def set(self, key: str, value: T) -> None:
        with self._lock:
            self._store[key] = (time.time() + self._ttl, value)

    def clear(self) -> None:
        with self._lock:
            self._store.clear()


class RateLimiter:
    """Fixed-window per-key rate limiter."""

    def __init__(self, limit: int, window_seconds: int) -> None:
        self._limit = limit
        self._window = window_seconds
        self._hits: dict[str, list[float]] = {}
        self._lock = threading.Lock()

    def allow(self, key: str) -> bool:
        now = time.time()
        with self._lock:
            hits = [t for t in self._hits.get(key, []) if now - t < self._window]
            if len(hits) >= self._limit:
                self._hits[key] = hits
                return False
            hits.append(now)
            self._hits[key] = hits
            return True

File: backend/src/cache.ts
/**
 * Tiny in-memory TTL cache and fixed-window rate limiter.
 *
 * Dependency-free. Sufficient for a single-instance deployment; swap for Redis
 * if the app is scaled horizontally. (Node is single-threaded per process, so
 * no explicit locking is needed.)
 */

export class TTLCache<T> {
  private ttlMs: number
  private store = new Map<string, { expiresAt: number; value: T }>()

  constructor(ttlSeconds: number) {
    this.ttlMs = ttlSeconds * 1000
  }

  get(key: string): T | null {
    const entry = this.store.get(key)
    if (!entry) return null
    if (Date.now() > entry.expiresAt) {
      this.store.delete(key)
      return null
    }
    return entry.value
  }

  set(key: string, value: T): void {
    this.store.set(key, { expiresAt: Date.now() + this.ttlMs, value })
  }

  clear(): void {
    this.store.clear()
  }
}

/** Fixed-window per-key rate limiter. */
export class RateLimiter {
  private limit: number
  private windowMs: number
  private hits = new Map<string, number[]>()

  constructor(limit: number, windowSeconds: number) {
    this.limit = limit
    this.windowMs = windowSeconds * 1000
  }

  allow(key: string): boolean {
    const now = Date.now()
    const recent = (this.hits.get(key) ?? []).filter((t) => now - t < this.windowMs)
    if (recent.length >= this.limit) {
      this.hits.set(key, recent)
      return false
    }
    recent.push(now)
    this.hits.set(key, recent)
    return true
  }
}
